### PR TITLE
add scrypt

### PR DESCRIPTION
**Pull request by nikgraf**

### CODE REVIEW

File: src/internal/passwordHash.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import * as Scrypt from "@noble/hashes/scrypt"
+import * as Data from "effect/Data"
+import * as Effect from "effect/Effect"
+
+class ScryptOptionsError extends Data.Error<{ message: string }> {}
```
Comment (fubhy) [inline src/internal/passwordHash.ts]: Just make it this:

```ts
{
  cause: unknown
}
```

Also, let's use `Data.TaggedError` (https://effect.website/docs/error-management/yieldable-errors#datataggederror)

File: src/internal/passwordHash.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+export const scrypt = (input: string | Uint8Array, salt: string | Uint8Array, options: Scrypt.ScryptOpts) =>
+  Effect.try({
+    try: () => Scrypt.scrypt(input, salt, options),
+    catch: (error) =>
```
Comment (fubhy) [inline src/internal/passwordHash.ts]:
```ts
catch: (cause) => new ScryptOptionsError({ cause })
```

File: src/index.ts [hunk 3]
```diff
@@ -51,3 +52,11 @@ export const sha512_224: (input: string | Uint8Array) => Uint8Array = hash.sha51
...
  */
 export const sha512_256: (input: string | Uint8Array) => Uint8Array = hash.sha512_256
+
+/**
+ * Password hashing the password and salt with scrypt.
+ *
+ * @since 1.0.0
+ * @category password hashing
+ */
+export const scrypt: typeof passwordHash.scrypt = passwordHash.scrypt
```
Comment (fubhy) [inline src/index.ts:62]: Let's avoid `typeof` here and "accept" the repetition. I'm also "team inference" but here, the intend is to expose it fully readable for lib consumers.

Alternatively, we can also ditch the `internal/*` modules (until we need them) and just put all the logic from those in-line right here.

File: src/internal/passwordHash.ts (hunk 1 above)
Comment (fubhy) [reply src/internal/passwordHash.ts]: Also, if we keep the `internal/*` modules, then this should just be named `OptionsError` (`Scrypt.OptionsError`).

**Review by fubhy — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 3 above)
Comment (nikgraf) [reply src/index.ts:62]: yeah, let's ditch the internal module. Writing out all the stuff is painful. But is typeof a problem? In VSCode it's showing the correct signature and now the typeof

Screenshot 2024-01-08 at 22 17 36

File: src/internal/passwordHash.ts (hunk 2 above)
Comment (nikgraf) [reply src/internal/passwordHash.ts:10]: Doesn't work this way?

Screenshot 2024-01-08 at 22 18 27

File: src/index.ts (hunk 3 above)
Comment (fubhy) [reply src/index.ts:62]: `typeof` is fine. The point here (in the entrypoints of the public modules as well as the root entrypoints) is that having the type signatures written out and paired with the documentation comments  makes these entrypoints extremely readable without having to further traverse into the underlying internals and/or using the IDE. That's all.

Comment (fubhy) [reply src/index.ts:62]: Ok then, let's ditch the internal modules (whose point really only is to circuit break circular dependencies), until we actually need them.

File: src/internal/passwordHash.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import * as Scrypt from "@noble/hashes/scrypt"
+import * as Data from "effect/Data"
+import * as Effect from "effect/Effect"
+
+class OptionsError extends Data.TaggedError("OptionsError")<{
+  cause: unknown
+}> {}
```
Comment (fubhy) [inline src/internal/passwordHash.ts:7]: Let's also put the `options` here? I'd probably not include the `input` and `salt` out of an abundance of caution to not accidentally expose any of those).

**Review by fubhy — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 3 above)
Comment (nikgraf) [reply src/index.ts:62]: but does this mean you would even copy & paste for example `ScryptOpts` which is internally coming from noble?

My main concern copying the types would be that if e.g. a new optional option gets added with a package upgrade you might miss adding it.

File: src/internal/passwordHash.ts (hunk 4 above)
Comment (nikgraf) [reply src/internal/passwordHash.ts:7]: It has pretty good error messages mentioning exactly what was wrong. Should we not focus in exposing these ones?

Comment (fubhy) [reply src/internal/passwordHash.ts:7]: Yeah we are exposing those through the `cause`. Do you think `options` wouldn't add any value? If that's the case then I'm happy to not include it.

File: src/index.ts (hunk 3 above)
Comment (fubhy) [reply src/index.ts:62]: For `options` structs no, I'd not repeat them here I think. It's a trade off between expressiveness (DX) and maintenance burden. I personally prefer the DX gain and readability of the lib code (which we also publish through `dist/src`).

File: src/internal/passwordHash.ts (hunk 4 above)
Comment (nikgraf) [reply src/internal/passwordHash.ts:7]: Don't think so. It's not like you pass in different value per user.

**Comment by fubhy** [comment]
@nikgraf As we flesh out & effectify the first few re-exports here we'll likely discuss more of these patterns. Once we've found a good set of rule of thumbs, things will get easier ;-)

**Comment by nikgraf** [comment]
yeah, that's why I always adding exactly one. Easier to figure these things :)

File: src/index.ts (hunk 3 above)
Comment (nikgraf) [reply src/index.ts:62]: That sounds like an useful trade-off :)

File: src/internal/passwordHash.ts (hunk 4 above)
Comment (nikgraf) [reply src/internal/passwordHash.ts:7]: To be honest, you shouldn't even use scrypt :D You should use argon2 - should we add that to the docs? basically https://www.latacora.com/blog/2018/04/03/cryptographic-right-answers/

Comment (fubhy) [reply src/internal/passwordHash.ts:7]: Yeah that makes sense

**Pull request merged**